### PR TITLE
RNA Connection: refactoring and simplification

### PR DESCRIPTION
**Pull request by sergeymitr**
#### Changes proposed in this Pull Request:
- add REST endpoint to get the authorization URL
- get rid of the `authorizationUrl` parameter, retrieve it from the REST API
- extract the user connecting functionality into the separate `ConnectUser` component

#### Jetpack product discussion
p1HpG7-bEn-p2

#### Does this pull request change what data or activity we track or use?
No.

#### Testing instructions:
0. Read the [README.md](https://github.com/Automattic/jetpack/pull/19837/files#diff-887d7021f0df39dd467af1107842df6c0340ebc25120890a2e5566b66060f64b) changes to see if the instructions are clear and correct, and there are no typos.
1. Disconnect Jetpack.
2. Go to "Tools -> Connection Manager" and click "Connect" to register the site. After a little wait, you'll see the In-Place iframe.
3. Confirm that the line "Site Registered" appears above the "Connect" button.
3. Do not approve the user connection, refresh the page.
4. Click "Connect" again to complete the user connection. The In-Place iframe should appear, click "Approve".
5. The "Connect" button should disappear, the page should inform you that you're fully connected.

**Please Note:** if you are testing this PR in Jetpack Monorepo, Connection Manager UI will get enabled automatically.
Otherwise, you need to create a plugin with the following code to enable it: https://gist.github.com/sergeymitr/bc005e1b893e778c03e4f43debb91c30

### CODE REVIEW

**Comment by sergeymitr** [comment]
Switching this back to "In Progress", as I decided to further refactor the component in this PR.

**Comment by sdixon194** [comment]
With in-place connections going away in #19872, is that something that will be removed for the RNA connection component as well, or will it continue to use the InPlaceConnection component even after it's been removed from the dashboard?

File: projects/packages/connection-ui/_inc/reducers/connection-status.js [hunk 1]
```diff
@@ -1,23 +1,12 @@
...
-				...state,
-				isRegistered: action.isRegistered,
-			};
-		case CONNECTION_STATUS_USER_CONNECTED:
-			return {
-				...state,
-				isUserConnected: action.isUserConnected,
-			};
+		case SET_CONNECTION_STATUS:
+			return action.connectionStatus;
```
Comment (samiff) [inline projects/packages/connection-ui/_inc/reducers/connection-status.js:9]: Is there a reason to completely overwrite the state object here versus using:
`return { ...state, ...action.connectionStatus };` ?

**Review by samiff — COMMENTED** [review]
This tests well for me 👍   I like that the consumer no longer needs to fetch the connection status.

File: projects/packages/connection-ui/_inc/reducers/connection-status.js (hunk 1 above)
Comment (sergeymitr) [reply projects/packages/connection-ui/_inc/reducers/connection-status.js:9]: I believe this makes the behavior more straightforward.
The spread construction you're suggesting is elegant, but will only merge top-level elements, while sub-object elements (e.g. children of `offlineMode`) will get overwritten.

So to get rid of such edge cases I decided to overwrite the object completely in the same way it's received from the RNA Connection. If there's a need to store some custom connection data, they should be stored separately.

**Comment by sergeymitr** [comment]
Hi @sdixon194.

> With in-place connections going away in #19872, is that something that will be removed for the RNA connection component as well, or will it continue to use the InPlaceConnection component even after it's been removed from the dashboard?

You're right, we probably need to remove it rather sooner than later.
I'll work on it today.

**Comment by sergeymitr** [comment]
I decided to remove In-Place connection in a separate PR.
Marking this one ready for review again.

**Review by samiff — DISMISSED** [review]
This looks good to merge to me 👍 

> I decided to remove In-Place connection in a separate PR.

That makes sense since there is already a fair bit going on with this PR. Though it might be good to double check what is going on in regards to: https://github.com/Automattic/jetpack/pull/19872#issuecomment-844472913

File: projects/js-packages/connection/components/connect-user/index.jsx [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+		//throw new Error( 'Connect User URL (Authorization URL) is missing' );
+	}
+
+	if ( forceCalypsoFlow ) {
+		redirectFunc(
+			authorizationUrl +
+				( from
+					? ( authorizationUrl.includes( '?' ) ? '&' : '?' ) + 'from=' + encodeURIComponent( from )
+					: '' )
+		);
```
Comment (jeherve) [inline projects/js-packages/connection/components/connect-user/index.jsx:73]: Maybe in future iterations we could rely on `@wordpress/url` for things like this?
https://developer.wordpress.org/block-editor/reference-guides/packages/packages-url/#addQueryArgs

File: projects/packages/connection/src/class-rest-connector.php [hunk 3]
```diff
@@ -142,6 +142,27 @@ public function __construct( Manager $connection ) {
...
 		);
+
+		// Get authorization URL.
+		register_rest_route(
+			'jetpack/v4',
+			'/connection/authorize_url',
+			array(
+				'methods'             => WP_REST_Server::READABLE,
+				'callback'            => array( $this, 'connection_authorize_url' ),
+				'permission_callback' => '__return_true',
```
Comment (jeherve) [inline projects/packages/connection/src/class-rest-connector.php:153]: Shouldn't we limit this to folks who have the `jetpack_connect` cap?

File: projects/js-packages/connection/components/connect-user/index.jsx (hunk 2 above)
Comment (sergeymitr) [reply projects/js-packages/connection/components/connect-user/index.jsx]: Good idea, fixed!
https://github.com/Automattic/jetpack/pull/19837/commits/92fa0570bc1b3e32b0469a4aabda77bd0a4ef066

File: projects/packages/connection/src/class-rest-connector.php (hunk 3 above)
Comment (sergeymitr) [reply projects/packages/connection/src/class-rest-connector.php]: Fixed, good catch!
https://github.com/Automattic/jetpack/pull/19837/commits/3fc4dcafbdb147e14e3d8f034f2aa04cdba75364

**Comment by sergeymitr** [comment]
Too late to merge in the release cycle.
Putting this on hold until Jetpack 9.8 is branched off.

**Review by zinigor — APPROVED** [review]
Works as described for me!

**Pull request merged**